hack/generator/pkg/codegen/pipeline_remove_embedded_resources.go

```diff
@@ -0,0 +1,681 @@
+	// Gather information about the associated types
+	associatedCountPerContext := make(map[string]int)
+	for associated := range associatedNames {
+		embeddedName, err := parseContextualTypeName(associated)
+		if err != nil {
+			return nil, err
+		}
+		associatedCountPerContext[embeddedName.context] = associatedCountPerContext[embeddedName.context] + 1
+	}
+
+	if len(associatedCountPerContext) != 1 {
+		return nil, nil
+	}
+
+	// If all updated names share the same context, the context is not adding any disambiguation value so we can remove it
+	renames := make(map[astmodel.TypeName]astmodel.TypeName)
+	for associated := range associatedNames {
+		embeddedName, err := parseContextualTypeName(associated)
+		if err != nil {
+			return nil, err
+		}
+		embeddedName.context = ""
+		renames[associated] = embeddedName.ToSimplifiedTypeName()
+	}
+
+	return renames, nil
+}
+
+func (r renamer) simplifyRemainingAssociatedNames(_ astmodel.TypeName, associatedNames astmodel.TypeNameSet) (map[astmodel.TypeName]astmodel.TypeName, error) {
+	// remove _0, which especially for the cases where there's only a single
+	// kind of usage will make the type name much clearer
+	renames := make(map[astmodel.TypeName]astmodel.TypeName)
+	for associated := range associatedNames {
+		embeddedName, err := parseContextualTypeName(associated)
+		if err != nil {
+			return nil, err
+		}
+
+		possibleRename := embeddedName.ToSimplifiedTypeName()
+		if !possibleRename.Equals(associated) {
+			renames[associated] = possibleRename
+		}
+	}
+
+	return renames, nil
+}
+
+func (r renamer) performRenames(
+	renames map[astmodel.TypeName]astmodel.TypeName,
+	flag astmodel.TypeFlag) (astmodel.Types, error) {
+
+	result := make(astmodel.Types)
+
+	renamingVisitor := astmodel.MakeTypeVisitor()
+	renamingVisitor.VisitTypeName = func(this *astmodel.TypeVisitor, it astmodel.TypeName, ctx interface{}) (astmodel.Type, error) {
+		if newName, ok := renames[it]; ok {
+			return astmodel.IdentityVisitOfTypeName(this, newName, ctx)
+		}
+		return astmodel.IdentityVisitOfTypeName(this, it, ctx)
+	}
+
+	for _, def := range r.types {
+		updatedDef, err := renamingVisitor.VisitDefinition(def, nil)
+		if err != nil {
+			return nil, err
+		}
+		// TODO: If we don't remove this, something is causing these types to not be emitted.
+		// TODO: Unsure what that is... should track it down
+		updatedType, err := flag.RemoveFrom(updatedDef.Type())
+		if err != nil {
+			return nil, err
+		}
+		result.Add(updatedDef.WithType(updatedType))
+	}
+
+	return result, nil
+}
+
+// simplifyTypeNames simplifies contextual type names if possible.
+func simplifyTypeNames(types astmodel.Types, flag astmodel.TypeFlag) (astmodel.Types, error) {
+	renames := make(map[astmodel.TypeName]astmodel.TypeName)
+
+	// Find all of the type names that have the flag we're interested in
+	updatedNames := make(map[astmodel.TypeName]astmodel.TypeNameSet)
```

Comment:
I made this change because I initially agreed that it might be cleaner, but as you say since most of the remapping functions need both the `TypeName` format as well as the `embeddedResourceTypeName` format, you end up having to convert in one direction or the other. The biggest awkward bit with using `embeddedResourceTypeNameSet` is that type doesn't exist and you end up having to write it and some of the helper methods to make things cleaner (`Single()` for example). Given that at the end of the day it doesn't actually save us any conversions, I'm not sure it's worth it.

I've stashed this change away (I did like 90% of it before I concluded it wasn't really any better) in case other disagree, but for now will leave this as is.